server/controllers/business/search.js

```diff
@@ -0,0 +1,82 @@
+const Joi = require("joi");
+const { STATUS_CODES } = require("http");
+const { isAPIKeyPresent, fetchImageByCompanyFree } = require("../../services");
+const { updateApiUsageCount, isApiUsageLimitExceed } = require("../../services/Subscriptions");
+const { fetchUserByApiKey } = require("../../services/Keys");
+const Images = require("../../models/Images");
+
+const getSearchQuerySchema = Joi.object({
+  domainKey: Joi.string()
+    .regex(/^[A-Za-z0-9&-/:.]+$/)
+    .required().messages({
+      "any.required": "domainKey is required",
+      "string.pattern.base": "Invalid domainKey",
+    }),
+  API_KEY: Joi.string().guid({ version: "uuidv4" }).required().messages({
+    "any.required": "API key is required"
+  }),
+});
+
+async function searchLogoController(req, res, next) {
+  try{
+    const { error, value } = getSearchQuerySchema.validate(req.query);
+    if (!!error) {
+      return res.status(422).json({
+        message: error.message,
+        statusCode: 422,
+        error: STATUS_CODES[422],
+      });
+    }
+
+    const { domainKey, API_KEY } = value;
+    let companyNameBeginsWith = domainKey.replace(/.+\/\/|www.|\..+/g, "").toUpperCase();
```

Comment:
This controller queries the db 5+n times (n: number of images for the domain provided) before responding to the user. `isAPIKeyPresent`, `fetchUserByApiKey`, `isApiUsageLimitExceed`, `Images.find`, `fetchImageByCompanyFree`, `updateApiUsageCount`.
The response time will be very high. Try reducing the number of db calls.

Hint: Don't verify things that you don't need to.

Reply:
Hi @amankumarsingh77 ,

Thank you for highlighting this issue. It was indeed a significant performance bottleneck in the Search API. Based on your suggestions, I’ve made the following key improvements:

- Reduced redundant database calls for API key and user verification, and updated the `fetchImageByCompanyFree` method under Image Services to avoid repeated checks for image existence in the database.
- With these optimizations, the database is now queried only three times per API call, instead of the previous 5+n calls.

Reply:
Here's the updated test case result and coverage:

![Screenshot 2024-09-20 212453](https://github.com/user-attachments/assets/de8b63c9-36c6-4557-9222-879e4488fd3c)

<img src="https://github.com/user-attachments/assets/e704f0ab-7421-4903-a11b-ecd939f17010" alt="test-coverage" width="500" height="500">